File: leetcode/c++/Majority_Element.cpp
class Solution {
public:
    int majorityElement(vector<int> &num) {
        int cand = num[0];
        int count = 1;
        
        for(int i = 1; i < num.size(); i++){
            if(count == 0) {
                cand = num[i];
                count = 1;
                continue;
            }
            
            if(num[i] == cand) count++;
            else count--;
        }
        if(count > 0) return cand;
    }
};

File: leetcode/java/MajorityElement.java
class Solution {
    public int majorityElement(int[] nums) {
        if(nums == null || nums.length == 0) throw new IllegalArgumentException("x");
        int elem = nums[0];
        int cnt = 1;
        for(int i = 1; i < nums.length; i++) {
            if(elem == nums[i]) cnt++;
            else{
                if(cnt > 0) cnt--;
                else {
                    elem = nums[i];
                    cnt++;
                }
            }
        }
        return elem;
    }
}
